### PR TITLE
Add the wrap-fun-decl option

### LINKED ISSUE #626: Configure strategies for breaking long functions declarations
There is 3 strategies for functions declarations:
- Fits on the same line
	```
	let ffffffffffffffffffffff aaaaaaaaaaaaaaaaaaaaaa bbbbbbbbbbbbbbbbbbbbbb = g
	```
- Arguments fit on their line
	```
	let ffffffffffffffffffffff
		aaaaaaaaaaaaaaaaaaaaaa bbbbbbbbbbbbbbbbbbbbbb cccccccccccccccccccccc =
	  g
	```
- Otherwise, break on each arguments
	```
	let ffffffffffffffffffffff
		aaaaaaaaaaaaaaaaaaaaaa
		bbbbbbbbbbbbbbbbbbbbbb
		cccccccccccccccccccccc
		dddddddddddddddddddddd =
	  g
	```

Add an option to disable the second strategie and using the third instead.

Same with type annotations:
```
let ffffffffffffffffffffff : aaaaaaaaaaaaaaaaaaaaaa -> bbbbbbbbbbbbbbbbbbbbbb = g

let ffffffffffffffffffffff
	: aaaaaaaaaaaaaaaaaaaaaa -> bbbbbbbbbbbbbbbbbbbbbb -> cccccccccccccccccccccc =
  g

let ffffffffffffffffffffff
	:  aaaaaaaaaaaaaaaaaaaaaa
	-> bbbbbbbbbbbbbbbbbbbbbb
	-> cccccccccccccccccccccc
	-> dddddddddddddddddddddd =
  g
```

### PR DESCRIPTION
**Pull request by Julow**
Add the `break-fun-decl={wrap|fit-or-vertical|smart}` option.
It replaces `wrap-fun-args` (which now only control function applications) for function and method declarations and soon function in signatures.

- `break-fun-decl=wrap` is the old behavior of `wrap-fun-args=true`
	```
	let ffffffffffffffffffff aaaaaaaaaaaaaaaaaaaaaa bbbbbbbbbbbbbbbbbbbbbb
	    cccccccccccccccccccccc =
	  g
	```

- `break-fun-decl=smart` is the old behavior of `wrap-fun-args=false`
	```
	let ffffffffffffffffffff
	    aaaaaaaaaaaaaaaaaaaaaa bbbbbbbbbbbbbbbbbbbbbb cccccccccccccccccccccc =
	  g
	```

- `break-fun-decl=fit-or-vertical` is a stricter behavior, which fixes #626
	```
	let ffffffffffffffffffff
	    aaaaaaaaaaaaaaaaaaaaaa
	    bbbbbbbbbbbbbbbbbbbbbb
	    cccccccccccccccccccccc =
	  g
	```

This is still a work in progress. Function signatures are not implemented and there is a few regressions.

### CODE REVIEW

**Comment by gpetiot** [comment]
Is it still in progress or is it ready to be reviewed?

**Comment by Julow** [comment]
It is ready for a review. I didn't implement it on function signatures yet, I can do it in an other PR.

File: test/passing/fun_decl.ml [hunk 1]
```diff
@@ -1,77 +1,32 @@
...
 let to_loc_trace
     ?(desc_of_source =
       fun source ->
         let callsite = Source.call_site source in
         Format.asprintf
           "return from %a"
           Typ.Procname.pp
-          (CallSite.pname callsite))
-    ?(source_should_nest = fun _ -> true)
+          (CallSite.pname callsite)) ?(source_should_nest = fun _ -> true)
```
Comment (gpetiot) [inline test/passing/fun_decl.ml]: This regression should be fixed, `source_should_nest` should be indented on the same level as `desc_of_source` and the other function's parameters.

**Comment by gpetiot** [comment]
Looks good to me, I was wondering if it would look better if the arguments were indented?
```ocaml
 let ffffffffffffffffffff aaaaaaaaaaaaaaaaaaaaaa bbbbbbbbbbbbbbbbbbbbbb
       cccccccccccccccccccccc =
   g

 let ffffffffffffffffffff
       aaaaaaaaaaaaaaaaaaaaaa bbbbbbbbbbbbbbbbbbbbbb cccccccccccccccccccccc =
   g

 let ffffffffffffffffffff
       aaaaaaaaaaaaaaaaaaaaaa
       bbbbbbbbbbbbbbbbbbbbbb
       cccccccccccccccccccccc =
   g
```

**Comment by jberdine** [comment]
I wonder if `break-fun-decl=smart` is necessary. Maybe we should consider removing it, or deprecating it with a warning and see if anyone complains.

**Pull request merged**